### PR TITLE
feat: add grpc health check and graceful grpc server shutdown

### LINKED ISSUE #378: Support a health check endpoint to use as readiness and liveness kubernetes probes
**Is your feature request related to a problem? Please describe.**
Lack of readiness / liveness probes, which is needed for a graceful server shutdown.

**Describe the solution you'd like**
I'd like to get a gRPC server for health check, and use it to set up the readiness and liveness probes when running in kubernetes.

**Describe alternatives you've considered**
Not sure if there is a good alternative.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by loranger2k**
**What type of PR is this?**
/kind feat

**What this PR does / Why we need it**:
1. Provides a gRPC health check endpoint that can be hooked up to kubernetes for managing traffic to the pods.
2. Provides a graceful server shutdown when kubernetes sends SIGTERM to the pods when scaling down or when doing rolling deployment.

**Which issue(s) this PR fixes**:

Closes #378 #379 

**Special notes for your reviewer**:
There is a tool that can be used to hook the health check endpoint to kubernetes since native gRPC endpoint are not supported before 1.23 (alpha), 1.24 (beta). See details here on how to package the executable to your service container: https://github.com/grpc-ecosystem/grpc-health-probe

### CODE REVIEW

File: internal/app/server/server.go [hunk 1]
```diff
@@ -16,37 +16,101 @@ package server
...
+	select {
+	case <-sigs:
+		break
+	case <-ctx.Done():
+		break
+	}
+	log.Warn("received shutdown signal")
+
+	// Must be called as early as possible to stop accepting connections
+	cancel()
```
Comment (yuryu) [inline internal/app/server/server.go:98]: Are we sure we want to cancel ongoing requests and transactions? I think "graceful shutdowns" usually means the server waits for any requests that are being processed within a reasonable timeouts. Apache and Nginx work like this IIRC.

File: cmd/server/main.go [hunk 2]
```diff
@@ -40,6 +41,6 @@ func main() {
 
 	ctx := context.Background()
 	if err := server.Run(ctx, "tcp", cfg); err != nil {
-		log.Fatalf("got error starting server: %v", err)
+		os.Exit(1)
```
Comment (yuryu) [inline cmd/server/main.go:44]: Is there any reason changed this? I think Fatal is still appropriate for server errors.

**Review by yuryu — CHANGES_REQUESTED** [review]
Can you split the PR into two respective parts? It doesn't look like they depend on each other. Overall I agree with the proposal.

I added comments, and also we need tests before merging.

File: internal/app/server/server.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 Google LLC
+// Copyright 2022 Google LLC
```
[the remark this answers is not in the document]
Comment (loranger2k) [reply internal/app/server/server.go:1]: Will revert.

File: internal/app/server/server.go [hunk 4]
```diff
@@ -16,37 +16,101 @@ package server
...
+	"golang.org/x/sync/errgroup"
 	"google.golang.org/grpc"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
 	"google.golang.org/grpc/reflection"
 
 	pb "github.com/googleforgames/open-saves/api"
 )
 
-// Run starts the Open Saves gRPC service.
+const serviceName = "Health"
```
[the remark this answers is not in the document]
Comment (loranger2k) [reply internal/app/server/server.go:36]: Good suggestion. Will change.

File: internal/app/server/server.go (hunk 1 above)
Comment (loranger2k) [reply internal/app/server/server.go:98]: My understanding of cancel is that it does simply stops the service from accepting new connection. The graceful server shutdown takes care of waiting for the current requests to finish before quitting.

File: internal/pkg/config/model.go [hunk 5]
```diff
@@ -22,6 +22,7 @@ const (
 	OpenSavesBucket  = "open_saves_bucket"
 	OpenSavesProject = "open_saves_project"
 	LogLevel         = "log_level"
+	HealthCheckPort  = "health_check_port"
```
[the remark this answers is not in the document]
Comment (loranger2k) [reply internal/pkg/config/model.go:25]: We have been using a different port to run the Kubernetes liveness and readiness probes in all of our other services. If you want to enforce using the same port for both interfaces, can the config simply be changed?

File: internal/app/server/server.go [hunk 6]
```diff
@@ -16,37 +16,101 @@ package server
...
+	signal.Notify(sigs, os.Interrupt, syscall.SIGTERM)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	g, ctx := errgroup.WithContext(ctx)
 
-	lis, err := net.Listen(network, cfg.ServerConfig.Address)
+	// gRPC Health Server
+	grpcHealthServer := grpc.NewServer()
```
[the remark this answers is not in the document]
Comment (loranger2k) [reply internal/app/server/server.go:50]: This is a practice that I have seen done in multiple Golang services. It comes handy when you start adding other capabilities such as prometheus scrapping capabilities or endpoint for pprof for activating / pulling profiler's metrics as new servers.
This has been tested in other services we use, which have both gRPC and http servers mixed in. The main advantage is that there is no server that owns the main thread per say. This streamline the construct of the block dealing with running different servers.

File: internal/app/server/server.go (hunk 1 above)
Comment (yuryu) [reply internal/app/server/server.go:98]: I think that's what `GracefulStop()` does. Can you write a test for this behavior?

File: cmd/server/main.go (hunk 2 above)
Comment (loranger2k) [reply cmd/server/main.go:44]: OK, will revert.

File: internal/app/server/server.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (yuryu) [reply internal/app/server/server.go:50]: Sorry it looks like it's the common pattern to handle signal handling in the main flow.

For using a single grpc server, please check this example: https://github.com/grpc/grpc-go/blob/master/examples/features/health/server/main.go

File: internal/app/server/server.go (hunk 1 above)
Comment (loranger2k) [reply internal/app/server/server.go:98]: The cancel is not needed / useful here according to this post with using errgroup (but it won't cause issues either): https://gist.github.com/akhenakh/38dbfea70dc36964e23acc19777f3869?permalink_comment_id=3139644#gistcomment-3139644 

For sure I will be testing this in our current set up with live traffic to make sure this is working as intended.

File: internal/pkg/config/model.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (yuryu) [reply internal/pkg/config/model.go:25]: No, that wouldn't work.

[the remark this answers is not in the document]
Comment (loranger2k) [reply internal/pkg/config/model.go:25]: I am going to keep the Ping interface within :6000, and expose the internal health check endpoint used by Kubernetes at :6001. We will eventually also have to support prometheus and pprof, and we do that in other services with using servers at different ports to split internal vs. external traffic.

File: internal/app/server/server.go (hunk 1 above)
Comment (loranger2k) [reply internal/app/server/server.go]: cancel() is not necessary at all. I have added the initialization of the health check serving status.

Comment (yuryu) [reply internal/app/server/server.go]: Your pull request seems based on the gist you put there. Do you have a license from the original author? I don't see the license in the gist that this was derived from.

**Comment by loranger2k** [comment]
Closed #381 as completed via #382 and #385.

**Pull request closed without merging**